### PR TITLE
[INTEL MKL]  Added mkl_einsum op

### PR DESCRIPTION
**Pull request by vishakha-nervana**
Added mkl_einsum op Implementation, In addition to   adding mkl_einsum, It also refactors mkl_batch_matmul op.

### CODE REVIEW

File: tensorflow/core/kernels/mkl/BUILD
Comment (kkimdev) [inline tensorflow/core/kernels/mkl/BUILD:63]: nit, indentation

File: tensorflow/core/kernels/mkl/BUILD
Comment (kkimdev) [inline tensorflow/core/kernels/mkl/BUILD:73]: nit, indentation

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h
Comment (kkimdev) [inline tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h:1]: nit: copyright and header `#ifndef` guard

**Comment by gbaned** [comment]
@vishakha-nervana  Can you please check @kkimdev's comments and keep us posted ? Thanks!

**Comment by kkimdev** [comment]
Also, please wait for @penpornk 's review.

File: tensorflow/core/kernels/mkl/BUILD
Comment (vishakha-nervana) [reply tensorflow/core/kernels/mkl/BUILD]: done

File: tensorflow/core/kernels/mkl/BUILD
Comment (vishakha-nervana) [reply tensorflow/core/kernels/mkl/BUILD:72]: done

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h
Comment (vishakha-nervana) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h:17]: done

**Comment by vishakha-nervana** [comment]
@penpornk Please review my PR.I have acknowledged changes.

File: tensorflow/core/kernels/mkl/mkl_einsum_op.cc
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_einsum_op.cc:28]: We should avoid having duplicate code. I normally don't want MKL code (`#ifdef INTEL_MKL`) interspersed with vanilla TF code, but I think I'm willing to trade that with having just one copy of code. Asking @ezhulenev for second opinion.

@ezhulenev This PR adds an MKL Einsum op. 

The op kernel is the same as the normal Einsum, except that it [calls MKLContractOperands](https://github.com/tensorflow/tensorflow/blob/8ab17489a4d4936404381b97b5c8135b7f20a9b3/tensorflow/core/kernels/mkl/mkl_einsum_op.cc#L182-L184) instead of [ContractOperands](https://github.com/tensorflow/tensorflow/blob/8ea9e8cfab457e73056347aa2366d748448a5745/tensorflow/core/kernels/linalg/einsum_op_impl.h#L639-L641). [MKLContractOperands](https://github.com/tensorflow/tensorflow/blob/8ab17489a4d4936404381b97b5c8135b7f20a9b3/tensorflow/core/kernels/mkl/mkl_einsum_op.cc#L39-L127) is also largely similar to [ContractOperands](https://github.com/tensorflow/tensorflow/blob/8ea9e8cfab457e73056347aa2366d748448a5745/tensorflow/core/kernels/linalg/einsum_op_impl.h#L546-L586), apart from the part that it calls DNNL batch matmul primitive.

I'm thinking we should just have one Einsum op, refactor the DNNL batch matmul part to a function, and call that function in `ContractOperands` under `#ifdef INTEL_MKL`. What do you think?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the delay!

Since the code of MKL Einsum is mostly the same as the vanilla Einsum op, I think we can just modify the vanilla Einsum op to call DNNL batch matmul when MKL is enabled. But I'd like to hear from @ezhulenev before we proceed. So please wait before you make modifications related to this part.

File: tensorflow/core/kernels/mkl/mkl_einsum_op.cc
Comment (ezhulenev) [reply tensorflow/core/kernels/mkl/mkl_einsum_op.cc:28]: I'd say it depends on the path of integrating mkl/oneDnn into TF that we'll choose. If we'll go with integrating MKL into "default" kernels, then it makes sense to provide different implementations of "helpers" and "functors" based on the compile flags (Eigen or MKL) and keep a single "Op" implementation.

Although current approach leads to a lot of code duplication I like that it's consistent with other MKL ops and that the code is completely separate.

I'd ask TF-Core owners what they think, maybe @rjpower can chime in.

Comment (gbaned) [reply tensorflow/core/kernels/mkl/mkl_einsum_op.cc:28]: @rjpower  Any update on this PR? Please. Thanks!

Comment (rjpower) [reply tensorflow/core/kernels/mkl/mkl_einsum_op.cc:28]: Having consistency with the other MKL ops seems appropriate, even if there's some extra duplication.

@penpornk / @ezhulenev if you feel it's appropriate, should we factor out the "helper" code for einsum into einsum-lib.h/cc and reuse it here?

Comment (penpornk) [reply tensorflow/core/kernels/mkl/mkl_einsum_op.cc:28]: Thank you, Russell! Refactoring the helper code out sounds good to me. I want both ops to get the same fixes and optimizations.

File: tensorflow/core/common_runtime/mkl_layout_pass_test.cc
Comment (penpornk) [inline tensorflow/core/common_runtime/mkl_layout_pass_test.cc:5145]: Nit: Can we add a space between `node/attr` and `{` and vertical-align the `value`s?
```suggestion
      "node { name: 'B' op: 'Float32InputList'"
      " attr { key: 'N'            value { i: 2 } }}"
      "node { name: 'C' op: 'Einsum'"
      " attr { key: 'equation'     value { s: '->' }}"
      " attr { key: 'N'            value { i: 2 } }"
      " attr { key: 'T'            value { type: DT_FLOAT } }"
```

File: tensorflow/core/graph/mkl_graph_util.h
Comment (penpornk) [inline tensorflow/core/graph/mkl_graph_util.h:155]: Nit: Can we keep `BatchMatMul` and `BatchMatmulV2` together and put `Einsum` after `BatchMatMulV2`?
```suggestion
       0 == name.compare("BatchMatMul") || 0 == name.compare("BatchMatMulV2") ||
       0 == name.compare("Einsum") || 0 == name.compare("MatMul") ||
```

File: tensorflow/core/ops/linalg_ops.cc
Comment (penpornk) [inline tensorflow/core/ops/linalg_ops.cc:676]: Please move this to a separate MKL file, e.g., a new file named `mkl_linalg_ops.cc`. If you don't expect to have more linalg ops in the near future, please feel free to put it in `mkl_nn_ops.cc`.

**Review by penpornk — CHANGES_REQUESTED** [review]
I have a few more comments now that we decide to keep the MKL op (refactoring the common code to a helper file).

**Comment by vishakha-nervana** [comment]
@penpornk I have made changes as per your request, please let me know if there's any AR for me regarding refactoring.

**Comment by gbaned** [comment]
@vishakha-nervana Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@vishakha-nervana Any update on this PR? Please. Thanks!

**Comment by vishakha-nervana** [comment]
> @vishakha-nervana Any update on this PR? Please. Thanks!

I have resolved all the conflicts long, ago @penpornk Is there any change I need to do in this branch?

File: tensorflow/core/BUILD
Comment (vishakha-nervana) [reply tensorflow/core/BUILD]: Done.

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h
Comment (vishakha-nervana) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h:109]: Done.

**Review by penpornk — APPROVED** [review]
I'm so sorry for the delay! Thank you very much again for the PR!

**Comment by gbaned** [comment]
@vishakha-nervana Can you please address Ubuntu Sanity errors? Thanks!

**Comment by vishakha-nervana** [comment]
> @vishakha-nervana Can you please address Ubuntu Sanity errors? Thanks!

@gbaned Done.

**Review by penpornk — APPROVED** [review]
Thank you!

**Pull request merged**
